### PR TITLE
chore(middleware): Adds appBaseUrl, removes redirect_uri

### PR DESCRIPTION
**Pull request by swiftone**
BREAKING CHANGE:

Configurations now require appBaseUrl.
redirect_uri is deprecated

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our [guidelines](/okta/okta-oidc-js/blob/master/CONTRIBUTING.md#commit)
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [X] Adding Tests
- [ ] Build related changes
- [ ] CI related changes
- [X] Documentation changes
- [ ] Other... Please describe:

## What is the current behavior?

The ExpressOIDC() constructor expects to be passed a `redirect_uri` param

The optional `routes` option has a `callback` param that can have a `defaultRedirect` option
Issue Number: N/A

## What is the new behavior?
The ExpressOIDC() constructor expects an `appBaseUrl` param (it will construct default `redirect_uri` from that and the `routes`, be those default or custom)
* `routes.callback` is now `routes.loginCallback`
* `routes.callback.defaultRedirect` is now `routes.loginCallback.afterCallback`

## Does this PR introduce a breaking change?
- [X] Yes
- [ ] No

See "Updating" in the README

## Other information

There will be a follow-up breaking change to add a new default `/logout` route

## Reviewers
@manueltanzi-okta @robertdamphousse-okta

### CODE REVIEW

File: packages/oidc-middleware/src/ExpressOIDC.js [hunk 1]
```diff
@@ -68,28 +71,35 @@ module.exports = class ExpressOIDC extends EventEmitter {
...
+    // Build redirect uri unless explicitly set
+    options.loginRedirectUri = loginRedirectUri || `${appBaseUrl}${options.routes.loginCallback.path}`;
+
+    // Validate the redirect_uri param
+    assertRedirectUri(options.loginRedirectUri);
+
     const context = {
-      options
+      options,
+      emitter: this
```
Comment (manueltanzi-okta) [inline packages/oidc-middleware/src/ExpressOIDC.js]: Curious, what's the reason for this change? To make it more explicit?

File: yarn.lock [hunk 2]
```diff
@@ -566,6 +566,11 @@
     pouchdb-collections "^1.0.1"
     tiny-queue "^0.2.1"
 
+"@okta/configuration-validation@^0.1.1":
+  version "0.1.1"
```
Comment (manueltanzi-okta) [inline yarn.lock]: Wonder, should this be 0.2.0?

File: packages/oidc-middleware/README.md [hunk 3]
```diff
@@ -139,10 +143,11 @@ Required config:
 * **issuer** - The OIDC provider (e.g. `https://{yourOktaDomain}/oauth2/default`)
 * **client_id** - An id provided when you create an OIDC app in your Okta Org
 * **client_secret** - A secret provided when you create an OIDC app in your Okta Org
-* **redirect_uri** - The callback for your app. Locally, this is usually `http://localhost:3000/authorization-code/callback`. When deployed, this should be `https://{yourProductionDomain}/authorization-code/callback`.
+* **appBaseUrl** - The base scheme, host, and port (if not 80/443) of your app, not including any path (e.g. http://localhost:3000, not http://localhost:3000/ )  You may specific `loginRedirectUri` to override this setting if you redirect to other apps.
```
Comment (manueltanzi-okta) [inline packages/oidc-middleware/README.md]: I think `specify` instead of `specific`.
nit: Also, not sure if we should say `to override this setting`, since they should have different roles. Maybe we can instead explain its standalone behavior, since the override is already explained in `loginRedirectUri`.

File: packages/oidc-middleware/README.md [hunk 4]
```diff
@@ -139,10 +143,11 @@ Required config:
 * **issuer** - The OIDC provider (e.g. `https://{yourOktaDomain}/oauth2/default`)
 * **client_id** - An id provided when you create an OIDC app in your Okta Org
```
Comment (manueltanzi-okta) [inline packages/oidc-middleware/README.md]: Let's leave this for the future, but probably at some point we should standardize to all camelCase or all underscore_case.

File: packages/oidc-middleware/README.md [hunk 5]
```diff
@@ -366,6 +376,20 @@ Once you have done that you can read the documentation on the [request][] librar
 [openid-client]: https://github.com/panva/node-openid-client
 [request]: https://github.com/request/request
 
+### Upgrading 
+
+#### from 1.x to 2.x
```
Comment (manueltanzi-okta) [inline packages/oidc-middleware/README.md]: Should we mention they need to add `appBaseUrl`?

File: yarn.lock (hunk 2 above)
Comment (swiftone) [reply yarn.lock]: It should...and I have no idea why it's not updating.

File: packages/oidc-middleware/src/ExpressOIDC.js (hunk 1 above)
Comment (swiftone) [reply packages/oidc-middleware/src/ExpressOIDC.js]: Currently, only the ExpressOIDC object or something with a reference to it can emit.  

Now that we can have errors on more than setup, (such as an error during token revoke) we need a way to emit errors on those, and those don't happen in ExpressOIDC, but rather in another call that ExpressOIDC calls.  

We could have THOSE be emitters and have ExpressOIDC listen and re-emit, or we can just pass the emitter down so those can call the emitter directly when needed.

File: packages/oidc-middleware/README.md (hunk 3 above)
Comment (swiftone) [reply packages/oidc-middleware/README.md]: Agreed

File: packages/oidc-middleware/README.md (hunk 4 above)
Comment (swiftone) [reply packages/oidc-middleware/README.md]: I think this is a transition in process

File: packages/oidc-middleware/README.md (hunk 5 above)
Comment (swiftone) [reply packages/oidc-middleware/README.md]: That's the first line under the next subheading

Comment (swiftone) [reply packages/oidc-middleware/README.md]: ...or I'm looking at the wrong version of the file.  I'll update to add adding appBaseUrl and removing redirect_uri
